Feature/#29 projects storage closes #29

sfiquet opened the pull request:
This implements the GET route for projects. 

It took me a while to figure out how to do testing. In the end I went with Eric Elliott's view that [routes don't need unit testing, they need integration testing](https://medium.com/javascript-scene/mocking-is-a-code-smell-944a70c90a6a). 

I wanted to separate unit tests (that only access JS files and should be fairly fast) from integration tests (that access a database so require more set-up and are slower) so I've added script `testint` specifically for integration testing. 

Then it turned out that running `npm test` normally would also run those tests but on the development database instead of the test database, so I fiddled with the database path and the `test` script. The idea behind the new implementation is that integration tests will be named `*.int.test.js` and unit tests will be `*.unit.test.js`. As it stands the server currently only has integration tests so maybe it wasn't necessary but at least the structure is there if we need it.

Heyjp commented on components/projects/projectmodel.js:
```diff
@@ -0,0 +1,20 @@
+const { Schema } = mongoose;
+
+const ProjectSchema = new Schema({
+  title: String,
```
Currently a project can be saved even if there is no information provided to the model. I think at least the `title` and the `authorId` should be required fields if a project is to be saved.

sfiquet replied:
It does? Oh, good catch, I'll look into that.

sfiquet replied:
I see what's going on. This issue doesn't deal with saving the project at all, it's all about the GET. The POST will be handled with issue #46 but I'll make the fields required now.

Heyjp commented:
Only thing I noticed was that `getAllProjects` whilst fine right now, if the app was to get a larger user base being able to get all the projects might be unnecessary and slow down the app. I see you placed the query error, so are parameters something that will be added in at a later date?

sfiquet commented on components/projects/projects.int.test.js:
```diff
@@ -0,0 +1,187 @@
+  await ProjectModel.remove();
+}
+
+async function setupProjects(projects) {
```
Yes, I was not doing the exact same thing on all of them, that's why I didn't use the standard setup and tear down. But I think I need to have another look anyway because the data should be removed from the database by the end of each test and I didn't do that. Maybe I'll put the test on an empty database in its own describe block or something and do a common setup for everything else.

sfiquet commented:
The thing that makes sense is to have a filter functionality for the user to look for projects that interest them, e.g. a particular language or domain. 

But your point about not returning the whole table if we have lots makes a lot of sense. Would you limit the number of projects returned then? What would be a good limit? Should that be set by the client? Or would we set up the server to page its results? Do you think we need to do this now for the MVP or add it as a further issue? I would tend to make it an additional issue.

sfiquet commented on components/projects/projectmodel.js:
```diff
@@ -0,0 +1,20 @@
+  authorName: String,
+  keywords: [String],
+  description: String,
+  lastSaved: Date,
```
Thanks I'll look into the timestamps.

As a user, what are you expecting from a date of creation for the project? I am not sure what value it adds.

I think it would actually be a date of first publication, because if the owner sits on the draft for a month or two before publishing, that's none of the other users' business.

So when you look at a project, would you see the date of last (published) change and the date of first publication? What are the use cases for needing this? I'm wondering if it's a case for separating the project from the help wanted ad or if you mean something different.

Heyjp replied:
I guess date of first published would do something similar, but possibly a project owner might want to look back and see how long their project has been running on the site. Maybe just for a sense of accomplishment more than anything if it's a long running thing. Also users may want to see when a project was made, if it was created a long time ago with not many updated they might not want to join.

But first published covers both of those.

Heyjp commented:
I think 10 projects is enough. One way I did it in a previous project was to return the amount of documents on the DB and then just map out a bunch of buttons which would send a request for the next set of docs up until all had been shown. So button 1 would show results 0-10, button 2,results 2-10 etc etc. 

Do you have any ideas on how to do this? But as long as it gets added as an issue I think that will be good enough.

sfiquet commented:
Implemented for the latest version:
- use Mongoose timestamps for dates
I also changed the implementation of draft and published so we now have separate objects for each version with its own creation date and last update date. This was inspired by [this blog post](http://software.danielwatrous.com/representing-revision-data-in-mongodb/).
- use Jest set-up and tear-down in tests and do a proper clean up
- make some fields required so that empty projects cannot be saved in the database

I created issue #72 for returning projects in batches/pages.
